--- parted-3.1/parted/ui.c.orig	2012-03-04 21:39:46.283904709 +0100
+++ parted-3.1/parted/ui.c	2012-03-06 15:55:54.862325698 +0100
@@ -161,6 +161,11 @@
 static const char* banner_msg = N_(
 "Welcome to GNU Parted! Type 'help' to view a list of commands.\n");
 
+static char* warning_msg = N_(
+"\nWARNING: unlike other partitioning software, Parted does *NOT* wait for\n"
+"exit to write changes to partition table, you're experimenting on your\n"
+"disk's live body.\n");
+
 static const char* usage_msg = N_(
 "Usage: parted [OPTION]... [DEVICE [COMMAND [PARAMETERS]...]...]\n"
 "Apply COMMANDs with PARAMETERS to DEVICE.  If no COMMAND(s) are given, "
@@ -1577,7 +1582,7 @@
 
         print_using_dev (*dev);
 
-        list = str_list_create (_(banner_msg), NULL);
+        list = str_list_create (_(banner_msg), _(warning_msg), NULL);
         str_list_print_wrap (list, screen_width (), 0, 0, stdout);
         str_list_destroy (list);
 
--- parted-3.2/po/pl.po.orig	2006-05-27 21:38:39.582850750 +0200
+++ parted-3.2/po/pl.po	2006-05-27 21:41:33.121696250 +0200
@@ -1717,6 +1717,18 @@
 
 #: parted/ui.c:71
 msgid ""
+"\n"
+"WARNING: unlike other partitioning software, Parted does *NOT* wait for\n"
+"exit to write changes to partition table, you're experimenting on your\n"
+"disk's live body.\n"
+msgstr ""
+"\n"
+"UWAGA: w przeciwieństwie do innych programów do partycjonowania, Parted\n"
+"*NIE* nie czeka na wyjście z programu z zapisaniem zmian w tablicy partycji\n"
+"- używając go eksperymentuje się na żywym organizmie dysku.\n"
+
+#: parted/ui.c:71
+msgid ""
 "Usage: parted [OPTION]... [DEVICE [COMMAND [PARAMETERS]...]...]\n"
 "Apply COMMANDs with PARAMETERS to DEVICE.  If no COMMAND(s) are given, run in\n"
 "interactive mode.\n"
