--- parted-1.4.11/libparted/disk_gpt.c~	Thu May  3 17:25:31 2001
+++ parted-1.4.11/libparted/disk_gpt.c	Fri May  4 12:52:39 2001
@@ -545,8 +545,8 @@
 	memset(&pmbr, 0, sizeof(pmbr));
 	pmbr.Signature = MSDOS_MBR_SIGNATURE;
 	pmbr.PartitionRecord[0].OSType = EFI_PMBR_OSTYPE_EFI_GPT;
-	pmbr.PartitionRecord[0].EndHead = 0xFF;
-	pmbr.PartitionRecord[0].EndSector = 0xFE;
+	pmbr.PartitionRecord[0].EndHead = 0xFE;
+	pmbr.PartitionRecord[0].EndSector = 0xFF;
 	pmbr.PartitionRecord[0].EndTrack = 0xFF;
 	pmbr.PartitionRecord[0].StartingLBA = 1;
 	pmbr.PartitionRecord[0].SizeInLBA = LastLBA(dev) + 1;
