--- parted-1.7.1/include/parted/device.h.orig	2006-05-27 22:20:04.522149000 +0200
+++ parted-1.7.1/include/parted/device.h	2006-05-27 22:25:19.965863750 +0200
@@ -135,8 +135,6 @@
 extern int ped_device_sync_fast (PedDevice* dev);
 extern PedSector ped_device_check (PedDevice* dev, void* buffer,
                                    PedSector start, PedSector count);
-extern PedConstraint* ped_device_get_constraint (PedDevice* dev);
-
 /* private stuff ;-) */
 
 extern void _ped_device_probe (const char* path);
--- parted-1.7.1/include/parted/constraint.h.orig	2006-05-25 19:28:43.000000000 +0200
+++ parted-1.7.1/include/parted/constraint.h	2006-05-27 22:25:20.497897000 +0200
@@ -93,5 +93,8 @@
 extern PedConstraint*
 ped_constraint_exact (const PedGeometry* geom);
 
+extern PedConstraint*
+ped_device_get_constraint (PedDevice* dev);
+
 #endif /* PED_CONSTRAINT_H_INCLUDED */
 
