--- parted-3.5/parted/ui.c.orig	2022-05-08 08:11:19.070886745 +0200
+++ parted-3.5/parted/ui.c	2022-05-08 08:17:50.045435321 +0200
@@ -222,7 +222,7 @@ screen_width ()
 {
         int    width = 0;
 
-        if (opt_script_mode || pretend_input_tty)
+        if (opt_script_mode || pretend_input_tty || !isatty (0))
                 return 32768;    /* no wrapping ;) */
 
 /* HACK: don't specify termcap separately - it'll annoy the users. */
