--- parted-1.4.14/libparted/disk_dos.c.wiget	Tue Apr 24 01:23:05 2001
+++ parted-1.4.14/libparted/disk_dos.c	Fri Jun 29 12:46:37 2001
@@ -511,6 +511,7 @@
 	old_s = disk->dev->sectors;
 
 	if (probe_real_geom (disk, raw_part, offset)) {
+#ifndef NO_BIOS_GEOMETRY_WARNING
 		if (ped_exception_throw (
 			PED_EXCEPTION_ERROR,
 			PED_EXCEPTION_IGNORE_CANCEL,
@@ -546,7 +547,7 @@
 			disk->dev->sectors = old_s;
 			return 0;
 		}
-
+#endif
 		return 1;
 	} else {
 		if (ped_exception_throw (
