--- parted-1.8.7/include/parted/natmath.h.orig	2007-04-18 17:12:10.000000000 -0400
+++ parted-1.8.7/include/parted/natmath.h	2007-06-04 10:03:37.374232283 -0400
@@ -87,10 +87,10 @@
 extern const PedAlignment* ped_alignment_any;
 extern const PedAlignment* ped_alignment_none;
 
-extern inline PedSector
+extern PedSector
 ped_div_round_up (PedSector numerator, PedSector divisor);
 
-extern inline PedSector
+extern PedSector
 ped_div_round_to_nearest (PedSector numerator, PedSector divisor);
 
 #endif /* PED_NATMATH_H_INCLUDED */
--- parted-1.8.7/libparted/cs/natmath.c.orig	2007-04-18 17:12:10.000000000 -0400
+++ parted-1.8.7/libparted/cs/natmath.c	2007-06-04 10:06:39.470610953 -0400
@@ -75,13 +75,13 @@
 	return sector - abs_mod (sector, grain_size);
 }
 
-inline PedSector
+PedSector
 ped_div_round_up (PedSector numerator, PedSector divisor)
 {
                 return (numerator + divisor - 1) / divisor;
 }
 
-inline PedSector
+PedSector
 ped_div_round_to_nearest (PedSector numerator, PedSector divisor)
 {
                 return (numerator + divisor/2) / divisor;
