--- parted-1.6.24/include/parted/device.h.orig	2005-08-15 20:25:26.794119776 +0200
+++ parted-1.6.24/include/parted/device.h	2005-08-15 20:26:32.862075920 +0200
@@ -36,7 +36,8 @@
  	PED_DEVICE_DASD		= 8,
  	PED_DEVICE_VIODASD	= 9,
         PED_DEVICE_SX8          = 10,
-	PED_DEVICE_UBD		= 11
+	PED_DEVICE_UBD		= 11,
+	PED_DEVICE_AOE          = 12
 } PedDeviceType;
 
 typedef struct _PedDevice PedDevice;
--- parted-1.6.24/libparted/linux.c.orig	2005-08-15 20:25:26.798119168 +0200
+++ parted-1.6.24/libparted/linux.c	2005-08-15 20:27:40.800747672 +0200
@@ -224,6 +224,7 @@
 
 #define DASD_MAJOR		94
 #define VIODASD_MAJOR		112
+#define AOE_MAJOR		152
 
 #define SCSI_BLK_MAJOR(M) (						\
 		(M) == SCSI_DISK0_MAJOR 				\
@@ -330,6 +331,8 @@
 		dev->type = PED_DEVICE_CPQARRAY;
 	} else if (dev_major == UBD_MAJOR && (dev_minor % 0x10 == 0)) {
 		dev->type = PED_DEVICE_UBD;
+	} else if (dev_major == AOE_MAJOR && (dev_minor % 0x10 == 0)) {
+		dev->type = PED_DEVICE_AOE;
 	} else {
 		dev->type = PED_DEVICE_UNKNOWN;
 	}
@@ -963,6 +966,11 @@
 			goto error_free_arch_specific;
 		break;
 
+	case PED_DEVICE_AOE:
+		if (!init_generic (dev, _("AoE Driver")))
+			goto error_free_dev;
+		break;
+
 	case PED_DEVICE_UBD:
 		if (!init_generic (dev, _("UBD Driver")))
 			goto error_free_dev;
