Only in parted-1.4.21-/parted: tags
diff -ur parted-1.4.21/parted/ui.c parted-1.4.21-/parted/ui.c
--- parted-1.4.21/parted/ui.c	Fri Dec 14 12:36:55 2001
+++ parted-1.4.21-/parted/ui.c	Fri Dec 14 12:36:38 2001
@@ -71,6 +71,11 @@
 "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the "
 "GNU General Public License for more details.\n");
 
+static char* warning_msg = N_(
+"\nWARNING: unlike other partitioning software, parted does *NOT* wait for\n"
+"exit to write changes to partition table, you're experimenting on your\n"
+"disk's live body.\n");
+
 static char* usage_msg = N_(
 "Usage: parted [OPTION]... [DEVICE [COMMAND [PARAMETERS]...]...]\n"
 "Apply COMMANDs with PARAMETERS to DEVICE.  If no COMMAND(s) are given, runs in\n"
@@ -643,7 +648,7 @@
 
 	printf ("%s", prog_name);
 
-	list = str_list_create (_(banner_msg), NULL);
+	list = str_list_create (_(banner_msg), _(warning_msg), NULL);
 	str_list_print_wrap (list, screen_width (), 0, 0);
 	str_list_destroy (list);
 
