diff -ur parted-1.4.16/libparted/device.c parted-1.4.16-/libparted/device.c
--- parted-1.4.16/libparted/device.c	Sat Aug 18 13:44:12 2001
+++ parted-1.4.16-/libparted/device.c	Sat Aug 18 13:44:44 2001
@@ -17,6 +17,8 @@
     Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */
 
+#define _GNU_SOURCE
+
 #include "config.h"
 
 #include <libintl.h>
@@ -890,7 +890,7 @@
 #endif
 	{
 		off_t		pos = sector * 512;
-		return lseek (dev->fd, pos, SEEK_SET) == pos;
+		return lseek64 (dev->fd, pos, SEEK_SET) == pos;
 	}
 }
 
