--- parted-1.7.1/configure.ac.orig	2005-03-19 22:47:12.000000000 +0100
+++ parted-1.7.1/configure.ac	2005-04-24 15:35:55.738042232 +0200
@@ -264,7 +265,7 @@
 if test "$with_readline" = yes; then
 	OLD_LIBS="$LIBS"
 	LIBS=""
-	AC_SEARCH_LIBS(tgetent, ncurses curses termcap termlib,
+	AC_SEARCH_LIBS(tgetent, tinfo ncurses curses termcap termlib,
 		PARTED_LIBS="$PARTED_LIBS $LIBS",
 		AC_MSG_ERROR(
 termcap could not be found which is required for the
