--- parted-1.7.1/include/parted/device.h.orig	2006-05-27 22:27:39.758600250 +0200
+++ parted-1.7.1/include/parted/device.h	2006-05-27 22:28:10.212503500 +0200
@@ -46,7 +46,8 @@
         PED_DEVICE_DASD         = 9,
         PED_DEVICE_VIODASD      = 10,
         PED_DEVICE_SX8          = 11,
-        PED_DEVICE_DM           = 12
+        PED_DEVICE_DM           = 12,
+	PED_DEVICE_AOE		= 13
 } PedDeviceType;
 
 typedef struct _PedDevice PedDevice;
--- parted-1.7.1/libparted/arch/linux.c.orig	2006-05-27 22:27:39.770601000 +0200
+++ parted-1.7.1/libparted/arch/linux.c	2006-05-27 22:29:40.138123500 +0200
@@ -238,6 +238,7 @@
 #define VIODASD_MAJOR           112
 #define SX8_MAJOR1              160
 #define SX8_MAJOR2              161
+#define AOE_MAJOR		152
 
 #define SCSI_BLK_MAJOR(M) (                                             \
                 (M) == SCSI_DISK0_MAJOR                                 \
@@ -438,6 +439,8 @@
                 dev->type = PED_DEVICE_CPQARRAY;
         } else if (dev_major == UBD_MAJOR && (dev_minor % 0x10 == 0)) {
                 dev->type = PED_DEVICE_UBD;
+	} else if (dev_major == AOE_MAJOR && (dev_minor % 0x10 == 0)) {
+		dev->type = PED_DEVICE_AOE;
         } else if (_is_dm_major(dev_major)) {
                 dev->type = PED_DEVICE_DM;
         } else {
@@ -1118,6 +1121,11 @@
                         goto error_free_dev;
                 break;
 
+	case PED_DEVICE_AOE:
+		if (!init_generic (dev, _("AoE Driver")))
+			goto error_free_dev;
+		break;
+
         case PED_DEVICE_FILE:
                 if (!init_file (dev))
                         goto error_free_arch_specific;
