Only in parted-1.4.19-/parted: tags
diff -ur parted-1.4.19/parted/ui.c parted-1.4.19-/parted/ui.c
--- parted-1.4.19/parted/ui.c	Tue Sep 18 23:16:37 2001
+++ parted-1.4.19-/parted/ui.c	Sun Oct 28 15:39:50 2001
@@ -91,6 +91,11 @@
 {
 	int	width = 0;
 
+	/* don't wrap lines if we are not writing to terminal
+	 * (crucial for scripting) */
+	if (!isatty(fileno(stdout)))
+		return 9999;
+
 /* HACK: don't specify termcap separately - it'll annoy the users. */
 #ifdef HAVE_LIBREADLINE
 	width = tgetnum ("co");
