--- parted-2.3/libparted/fs/reiserfs/reiserfs.c.orig	2010-02-08 07:48:18.000000000 +0100
+++ parted-2.3/libparted/fs/reiserfs/reiserfs.c	2010-09-04 09:30:07.798522522 +0200
@@ -726,13 +726,13 @@
 
 static int reiserfs_ops_init(void)
 {
-	if (!(libreiserfs_handle = dlopen("libreiserfs.so", RTLD_NOW)))
+	if (!(libreiserfs_handle = dlopen("libreiserfs-0.3.so.0", RTLD_NOW)))
 		goto error;
 
 	if (!reiserfs_ops_interface_version_check())
 		goto error_free_libreiserfs_handle;
 
-	if (!(libdal_handle = dlopen("libdal.so", RTLD_NOW)))
+	if (!(libdal_handle = dlopen("libdal-0.3.so.0", RTLD_NOW)))
 		goto error_free_libreiserfs_handle;
 
 	INIT_SYM(reiserfs_fs_probe);
